tooling/templatize/pkg/config/config.go

```diff
@@ -0,0 +1,191 @@
+		RegionShort: regionShort,
+		Stamp:       stamp,
+	}
+}
+
+type ConfigEvaluationContext struct {
+	Region      string
+	RegionShort string
+	Stamp       string
+}
+
+func (c *ConfigEvaluationContext) AsMap() map[string]interface{} {
+	return map[string]interface{}{
+		"ctx": map[string]interface{}{
+			"region":      c.Region,
+			"regionShort": c.RegionShort,
+			"stamp":       c.Stamp,
+		},
+	}
+}
+
+type ConfigProvider interface {
+	Validate(cloud, deployEnv string) error
+	GetVariables(cloud, deployEnv, region string, configEvalContext *ConfigEvaluationContext) (Variables, error)
+	GetDeployEnvVariables(cloud, deployEnv, region string, configEvalContext *ConfigEvaluationContext) (Variables, error)
+	GetRegions(cloud, deployEnv string) ([]string, error)
+	GetRegionOverrides(cloud, deployEnv, region string, configEvalContext *ConfigEvaluationContext) (Variables, error)
+}
+
+func NewConfigProvider(config string) ConfigProvider {
+	return &configProviderImpl{
+		config: config,
+	}
+}
+
+func interfaceToVariables(i interface{}) (Variables, bool) {
+	// Helper, that reduces need for reflection calls, i.e. MapIndex
+	// from: https://github.com/peterbourgon/mergemap/blob/master/mergemap.go
+	value := reflect.ValueOf(i)
+	if value.Kind() == reflect.Map {
+		m := Variables{}
+		for _, k := range value.MapKeys() {
+			m[k.String()] = value.MapIndex(k).Interface()
+		}
+		return m, true
+	}
+	return Variables{}, false
+}
+
+// Merges variables, returns merged variables
+// However the return value is only used for recursive updates on the map
+// The actual merged variables are updated in the base map
+func mergeVariables(base, override Variables) Variables {
+	for k, newValue := range override {
+		if baseValue, exists := base[k]; exists {
+			srcMap, srcMapOk := interfaceToVariables(newValue)
+			dstMap, dstMapOk := interfaceToVariables(baseValue)
+			if srcMapOk && dstMapOk {
+				newValue = mergeVariables(dstMap, srcMap)
+			}
+		}
+		base[k] = newValue
+	}
+
+	return base
+}
+
+func (cp *configProviderImpl) GetVariables(cloud, deployEnv, region string, configEvalContext *ConfigEvaluationContext) (Variables, error) {
+	variables, err := cp.GetDeployEnvVariables(cloud, deployEnv, region, configEvalContext)
+	if err != nil {
+		return nil, err
+	}
+
+	// region overrides
+	regionOverrides, err := cp.GetRegionOverrides(cloud, deployEnv, region, configEvalContext)
+	if err != nil {
+		return nil, err
+	}
+	mergeVariables(variables, regionOverrides)
+
+	return variables, nil
+}
+
+func (cp *configProviderImpl) Validate(cloud, deployEnv string) error {
+	config, err := cp.loadConfig(DefaultConfigEvaluationContext())
+	if err != nil {
+		return err
+	}
+	if _, ok := config.Overrides[cloud]; !ok {
+		return fmt.Errorf("the cloud %s is not found in %s", cloud, cp.config)
+	}
+
+	if _, ok := config.Overrides[cloud].Overrides[deployEnv]; !ok {
+		return fmt.Errorf("the deployment env %s is not found under cloud %s in %s", deployEnv, cloud, cp.config)
```

Comment:
What does the last `%s` do with `cp.config`?

Reply:
cp.config is the configfile path.
i removed it since it is not really necessary that the config file name is part of the error message for this particular situation